### DISCUSSION TITLE
Relicensing to BUSL-1.1

### DISCUSSION THREAD

**Discussion by kdy1** [dudykr/stc#304] [Polls]
Context:

I invested an enormous amount of time for stc, and I need to use more time for it.
And I want to be compensated for my hard work.
Also, I want to compensate contributors, too.

---

https://spdx.org/licenses/BUSL-1.1.html

BUSL has some parameters.

> The license parameters are: restrictions on usage, a change date, and the open source license that will govern usage of the software after the change date.

# License restrictions

## Non-commercial usage

You can use it, but dependants of your library/program should conform to the commercial license rule below if the dependant program/library is commercial.

## Commercial usage

Similar to [replicache](https://replicache.dev/#pricing), it will be free for companies with < $100k revenue (ARR).

If your company is bigger, you must sponsor stc via open collective.

See https://opencollective.com/stc

|   ARR  |   Tier you need  | Description |
|:------:|:----------------:|:-----------:|
| < 100K |        N/A       |     Free    |
|  < 1M  |  Bronze sponsor  |    500$/M   |
|  < 10M |  Silver sponsor  |   1000$/M   |
|  < 100M |   Gold sponsor   |   2500$/M   |
|  Extra | Platinum sponsor |   10000$/M  |

(Tier requirement for < 1M will be lowered in future)

# Change date and OSS license

It will be AGPL-3.0 after four years. (I may change this again in the future)

---

P.S.

I want to hear more opinions about the way to compensate contributors, too.

**Comment by castarco** [comment]
I think the tier between 100k and 1M is excessive, taking into account that the prices are based on revenue instead of based on profit, and would affect companies with revenue of 100k+1.

After 1M it starts to feel more acceptable.

**Reply by kdy1** [reply]
Thank you for your opinion!
I agree with your point. Since the vote is going on, I'll create a new vote for it after making decision.

**Comment by SaadBazaz** [comment]
In what capacity should they be paying this?
Let's say they just use it as a CI, which runs once every week. Do they have to pay the same as a company which uses STC to develop an entirely new framework and sell it as a PaaS? (Looking at Vercel)

**Reply by kdy1** [reply]
Yes, but maybe I should add more clauses for e.g. PAAS provides

**Reply by SaadBazaz** [reply]
Yeah, might be good to define usage and/or dependency, somehow.

**Comment by lukeapage** [comment]
At these tiers I don’t know if I could persuade where I work to pay for it.

Eg benefit is a few minutes off the build time, slightly better developer experience 

cost is 100 dollars a month per developer - we have a lot of outsourced developers in India for cost reduction- for 1 small tool we are now paying a percentage of 1 developers wage.

that’s more than say the cost of visual studio or other services/tools. If every part of the tool chain cost this much it would be extremely expensive.

maybe in America where a developer in some regions costs 400k it’s a drop in the ocean to pay a little extra for tooling.

I know it’s difficult to come up with pricing tiers and I sympathize that you want recompense for your efforts but I suspect these numbers will cause either your tool to not get much traction or companies to use it illegally and not pay you.

**Reply by lukeapage** [reply]
Also consider that some companies have bigger margins and some companies might have a small percentage of their workforce being frontenders ;)

**Comment by PedroHase** [comment]
I would argue to rethink the licensing. For small companies the current proposed licensing model would mean that they would essentially be barred from using it. As an example, I work for a small-ish company which would fall somewhere in between bronze & silver sponsor bracket, but could never afford (high revenue, high cost) nor justify paying $6k - 12k/year for a tool that would only be used by less than a handful of developers and only marginally increase productivity. Generally, I believe if a company's core competency isn't depending on the build speed like for PaaS providers (or if having long build times induces large costs), it's hard to justify the cost for the license in general, besides the price being extremely high for companies outside the US like @lukeapage pointed out.

At the same time I worry that bigger companies could use their subsidiaries to circumcise the licensing. Also what about startups that get pumped with millions of dollars, but won't see any meaningful revenue for a long time (maybe add a clause for funding amount)?

Personally, I'd love if the project could keep the current license and find a few corporate sponsors (e.g. PaaS provides like Vercel or Cloudflare). This way it would not bar devs/companies from adopting it, and have no legal implications which (hopefully) leads to a greater adoption of the tool. Alternatively, maybe additional services / capabilities could be offered which would require licensing or payment, but basic functionality would be open? 

In any case, looking forward, where the project goes and I hope you find a good solution :)

**Reply by kdy1** [reply]
Where does $6K come from? It's per-company not per-developer

**Reply by castarco** [reply]
I imagine he multiplied 500$ * 12 months.

**Reply by PedroHase** [reply]
> I imagine he multiplied 500$ * 12 months.

Indeed, I erroneously assumed that the "M" in "$500/M" stands for month. 

However, assuming that price is per million $ in yearly revenue, I believe my point would still stand that for smaller companies this would be difficult to bear and or justify (also see the point of @itaylor, most companies are not huge software companies where this would make a big enough impact )

**Reply by kdy1** [reply]
M is Month, I just missed `/Y`

**Comment by itaylor** [comment]
I understand your desire to find a good way to fund your work, but I think this license will not work out well.  Most companies are not huge software companies.  They have teams that write some software as part of running their business, making their website, etc.   Those teams will not be able to justify paying thousands of dollars a month in order to save a few seconds on their type checking vs tsc.  

For the huge tech companies, this license will also likely not work.  Large tech companies have procurement processes that are very time-consuming when negotiating a license that spans the whole company.   Teams within the large companies will avoid using stc to avoid going through the painful corporate licensing process.  Tying the license to the company's revenue will always force the license negotiation to happen at the corporate level, instead of at the team level.  IMO if your target is to sell subscriptions to large tech companies, you're better off making the licensing something that can happen at a per-team level.  X dollars for Y number of developers, or something else that doesn't get a corporate procurement/legal team involved to negotiate the price.  

You're kind of in a strange place here, as, by design, there is already a free alternative.  Trying to compete with a free product with one that has substantial restrictions will be difficult.

**Reply by bcheidemann** [reply]
I agree, per developer would be the way to go in my opinion.

**Reply by matthew-dean** [reply]
Per developer probably wouldn't work for most companies.

**Comment by kdy1** [comment]
Ah I see. Thanks

2022년 11월 17일 (목) 오후 4:30, Andrés Correa Casablanca <
***@***.***>님이 작성:

**Comment by ghost91-** [comment]
I agree with what some others have said: using a license model like this would probably mostly just cripple the usage of stc. Not just in corporate environments, but also in open source, because the suggested license is incompatible with other (permissive) open source licenses, so people will have reservations against using it, from that perspective.

I realize that you are investing a huge amount of time here, and that funding is a problem, but in my opinion, this is not the right way. I think a better solution would be to try to get some corporate sponsors to provide the necessary funds. It may be difficult at this point in time, where stc is not yet in a usable state, but this is how many successful open source projects work, and I believe it can work for stc, too.

**Comment by SaadBazaz** [comment]
I gave this some thought, and I think the best thing would be for you to provide your services for customizing PaaS platforms with bespoke solutions. 

That way you could have an array of "specialized" STCs for various domains (iygwim). Those you can definitely sell at a very good price, because any company looking for that can pay the "premium".

**Comment by 0xWOF** [comment]
In my opinion, If I can purchase or subscribe stc as personally and use it for personal project and company project, It would be great. Like IDE or Text Editor. Because purchase or subscribe as personally is easy, but pursuade company is difficult.

**Reply by bcheidemann** [reply]
@kdy1 I don't know if I'm in the minority here but I am always happy to pay for tools which make my life easier provided they won't break the bank. For fast type checking, since TS is my main language I use for work I would happily pay $5-$10 per month. I think personal and business licenses would not be mutually exclusive either.

But I guess it's not feasible if you don't want to make this into a company? 🤔

**Comment by advancedtw** [comment]
have a look at the recent discussion about mold linker and the upcoming license change. 
the tool saves a few seconds during dev, people like to use it but have warned the author that their company won't likely spend the extra cash when the slower but more established alterative is 100% free. 

also the poll will probably mislead you into thinking that the people who voted in favor will pay for it.  they could be under 100k ARR and simply not feel concerned or overestimating their company readiness to purchase a license so that devs can have a faster compile time.
It will be a harder sell when TS 5 lands and is promised to be faster by 10% or 20% (I haven't bench it just reporting what I've heard this information in speculative.)
as it stands a company making 99M pays 2500USD while a bootstrapped company making 120K per year has to pay 500USD. unless I'm missing something these numbers don't seem to scale that well. for that reason alone I recommend to not take the poll results too seriously.

stc will greatly benefit companies that want to safely compile very large code base or companies that compile a lot of code like PAAS. I think these are your paying customers. they can for example compare the cost of energy saving during ci builds and the cost of stc licensing and make an informed decision based on actual data.

**Comment by re-thc** [comment]
Sounds difficult to govern by revenue/profit as for a lot of companies this data isn't public. How would you measure it? Might be easier by the number of developers or some other metric.

Would it be possible to provide more value add if this will be paid, e.g. IDE integration or something else?
